File: JarvisServer/Services/ClientResponseTracker.cs
using System.Collections.Concurrent;

namespace JarvisServer.Services;

public class ClientResponseTracker
{
    private readonly ConcurrentDictionary<string, TaskCompletionSource<string>> messages = new();

    public (string Id, Task<string> Task) Register()
    {
        var correlationId = Guid.NewGuid().ToString();
        var tcs = new TaskCompletionSource<string>();

        var cts = new CancellationTokenSource(60000);
        cts.Token.Register(() =>
        {
            if (messages.TryRemove(correlationId, out var source))
            {
                source.TrySetCanceled();
            }
        });
        
        messages.TryAdd(correlationId, tcs);
        return (correlationId, tcs.Task);
    }

    public void Complete(string correlationId, string result)
    {
        if (messages.TryRemove(correlationId, out var tcs))
        {
            tcs.SetResult(result);
        }
    }
}

File: Server/Services/ClientResponseTracker.fs
namespace Server.Services

open System
open System.Collections.Concurrent
open System.Threading
open System.Threading.Tasks

type ClientResponseTracker() =
    let messages = ConcurrentDictionary<string, TaskCompletionSource<string>>()

    member this.Register() =
        let correlationId = Guid.NewGuid().ToString()
        let tcs = TaskCompletionSource<string>()

        let cts = new CancellationTokenSource(60000)

        cts.Token.Register(fun () ->
            match messages.TryRemove(correlationId) with
            | true, source -> source.TrySetCanceled() |> ignore
            | false, _ -> ())
        |> ignore

        messages.TryAdd(correlationId, tcs) |> ignore
        (correlationId, tcs.Task)

    member this.Complete(correlationId: string, result) =
        match messages.TryRemove(correlationId) with
        | true, source -> source.SetResult result
        | _ -> ()
